client/question/services/ConversationManagerService.js

```diff
@@ -13,18 +13,48 @@
 // limitations under the License.
 
 /**
- * @fileoverview Service that orchestrates the processing of a user's code
- * submission.
+ * @fileoverview Service that orchestrates TIE's conversation with the learner.
  */
 
-tie.factory('SolutionHandlerService', [
+tie.factory('ConversationManagerService', [
   '$q', 'CodePreprocessorDispatcherService', 'CodeRunnerDispatcherService',
   'FeedbackGeneratorService', 'PrereqCheckDispatcherService',
-  'TranscriptService', 'CodeSubmissionObjectFactory',
+  'TranscriptService', 'CodeSubmissionObjectFactory', 'FEEDBACK_CATEGORIES',
+  'FeedbackDetailsObjectFactory',
   function(
       $q, CodePreprocessorDispatcherService, CodeRunnerDispatcherService,
       FeedbackGeneratorService, PrereqCheckDispatcherService,
-      TranscriptService, CodeSubmissionObjectFactory) {
+      TranscriptService, CodeSubmissionObjectFactory, FEEDBACK_CATEGORIES,
+      FeedbackDetailsObjectFactory) {
+
+    // TODO(sll): Modify this to also depend on the learner's submission
+    // history.
+
+    /**
+     * Determines the feedback details that fully characterize the feedback
+     * that the learner should receive.
+     *
+     * @returns {FeedbackDetails}
+     */
+    var computeFeedbackDetails = function(codeEvalResult) {
+      var errorString = codeEvalResult.getErrorString();
+      if (errorString) {
+        if (errorString.startsWith('TimeLimitError')) {
+          return FeedbackDetailsObjectFactory.createTimeLimitErrorFeedback();
+        } else if (
+            errorString.startsWith('ExternalError: RangeError') ||
+            errorString.includes('maximum recursion depth exceeded')) {
```

Comment:
Given how we use these in test, as well, should these string literals be constants?

Reply:
I thought about this, and I don't think so. If we were to use them in other places in the (non-test) _code_, then making them constants makes sense for maintainability. But these are really used in just one place, and breaking them out into constants just moves the values far away from the code that uses them.

I'm fine with having to change the tests manually if the signature we're looking for ends up changing for some reason. It ensures that such a change is done intentionally.

Reply:
I'm fine with that; just wanted to make sure it was intentional. :)